### ISSUE TITLE
`getSession` fails during build when used in middleware

### ISSUE THREAD

**Issue by saevarb** [nextauthjs/next-auth#4467]
### Environment

react 17.0.2
next-auth 4.3.3
next 12.1.5

### Reproduction URL

N/A

### Describe the issue

`getSession` does not work in nextjs middlewares. Or rather, it works during development but fails when trying to build the project, with the following trace:

```
./node_modules/@babel/runtime/helpers/isNativeFunction.js
Dynamic Code Evaluation (e. g. 'eval', 'new Function') not allowed in Middleware pages/planning/_middleware

Import trace for requested module:
./node_modules/@babel/runtime/helpers/wrapNativeSuper.js
./node_modules/next-auth/core/errors.js
./node_modules/next-auth/lib/logger.js
./node_modules/next-auth/react/index.js
./src/server/auth_middleware.ts
./src/pages/overview/_middleware.ts
```

This seems to be a "known" issue, #4302, but that issue was just closed out of hand, even though it was "supposed to work". 

According to the documentation, we are supposed to use `withAuth` but this requires the JWT session strategy, which I'm not using. 

What is the recommended approach here? 

### How to reproduce

1. Create new next project with next-auth and session auth and any adapter
2. Try to use `getSession` in middlewares
3. Try to build the project

### Expected behavior

It doesn't fail to build.

**Comment by balazsorban44** [comment]
`getSession` should not be used in a middleware, please check out https://next-auth.js.org/configuration/nextjs#middleware

**Comment by saevarb** [comment]
@balazsorban44 
> `getSession` should not be used in a middleware, please check out https://next-auth.js.org/configuration/nextjs#middleware

Yes, I am aware of this and I mentioned in my issue, and explained why I can't use it(I'm not using JWT sessions). Is there a reason you ignored that?

Also, in the referenced issue you yourself say, quote:
> For a while, getSession has been our recommended method, so I suspect there is a config error or something

And this was only 24 days ago. So, if `getSession` can't be used in middleware, how can it have been the recommended method? 

Furthermore, I'd really like to know about any alternative approaches since I'm not using the JWT strategy. Can you confirm this means I just can't use middlewares until `getServerSession` lands or something like that?

**Comment by balazsorban44** [comment]
Sorry, the reason is that getSession is not compatible with the Edge Runtime, but we are working on it.

This is mentioned under caveats of the Middleware documentation

https://next-auth.js.org/configuration/nextjs#caveats

**Comment by hussamkhatib** [comment]
@saevarb I'm also not using jwt, what did you end up using

**Comment by HenriqueRamos13** [comment]
So, for who is using a non-JWT strategy, we need just wait getSession get compatible with edge functions? Or have another way?

**Comment by j-red** [comment]
It doesn't seem like a proper solution, but I have found a workaround for this (with Next v`13.5.3` and Next-Auth v`4.23.1`) -- `getServerSession()` wasn't working for me, so I'm using `getSession()` inside the middleware with the `'nodejs'` runtime (instead of `'edge'`) and have enabbled the unstable dynamic code execution for the specific problem file (in my case, `/node_modules/@babel/runtime/regenerator/index.js`). 

In my middleware file (`./middleware.tsx`):

```ts
export const config = {
  runtime: 'nodejs', // rather than "edge"
  unstable_allowDynamic: [
    '/node_modules/@babel/runtime/regenerator/index.js', // file causing the build error
  ],
};

export async function middleware(req: NextRequest) {
  const requestForNextAuth: Object = {
    headers: {
      cookie: req.headers.get('cookie'),
    },
  };

  const session = await getSession({ req: requestForNextAuth });

  // ... session checking logic

  return NextResponse.next();
}
```

I'm still relatively new to Next and Next-Auth so this may not be the best solution, but has addressed the build errors and enables me to deploy the application for now.

**Comment by DannelCu** [comment]
> Sorry, the reason is that getSession is not compatible with the Edge Runtime, but we are working on it.
> 
> The reason is that some of the core relies in Node.js, plus there isn't many Edge compatible database solutions yet, as far as I know.
> 
> This is briefly mentioned under caveats of the Middleware documentation.
> 
> https://next-auth.js.org/configuration/nextjs#caveats

can you as an expert confirm if @j-red solution is viable it worked for me but i do not know if whe are to face new problems in the future because of this.
